### ISSUE TITLE
Product variants should have unique SKUs (OCC-23)

### ISSUE THREAD

**Issue by DAud-IcI** [OrchardCMS/OrchardCore.Commerce#135]
Suggested by @chinasqzl [here](https://github.com/OrchardCMS/OrchardCore.Commerce/issues/5#issuecomment-1016995456).

This could be handled similarly to variant prices. However inventory management may complicate things. An approach where variants are separate content items and they are linked (similar to how the Content Localization feature works) may be more appropriate. Please discuss.

**Comment by DAud-IcI** [comment]
I suggest the following:
Product variant SKUs should be derived. For example if you a product with a base SKU of "SHIRT", and it has two sets of properties (size being S, M and L; color being black and blue) then you can use `productPart.Sku + '-' + value.Classify().ToUpperInvariant` to generate the SKU fragments. This was you'd have SKUs like "SHIRT-M-BLUE" or "SHIRT-L-BLACK".
At this point it would be good to convert the base SKU to upper-case during publish just to be clean since it's case-insensitive in the database anyway. Also now dash is an invalid character in the base SKU so that has to be handled.
Then we'd need to update `ProductService`, `ProductServiceExtensions` (basically just shop off the parts of the SKU variable after the first dash). Also add a `string GetVariantKey(string sku)` method and a `(PriceVariantsPart Part, string VariantKey) GetExactVariant(string sku)` that uses the full generated SKU so we can locate the variant, not just the product.

**Comment by porgabi** [comment]
> This way you'd have SKUs like "SHIRT-M-BLUE" or "SHIRT-L-BLACK".

This part has me a little confused at the moment. `PriceVariantsProduct` content items come with a singular `ProductPart` that has an `SKU` string field on it, but here the variants have no additional data associated with them. So there can be 3 different SKUs for just one of these items, stored on the `PriceVariantsPart` perhaps? It's unclear to me what the use case is going to be and how this should work in general.

Apart from general clarification, a few more specific questions:

1. Where should the variant-specific SKUs be saved? 
2. Should a base SKU still exist for `PriceVariantsProduct` items? 
3. Should these be editable/shown in the dashboard editor?

**Comment by 0liver** [comment]
From what I remember from our planning meeting, the variant SKUs are **transient**, they are not to be persisted anywhere. We generate them in code on the fly, and interpret them on the fly when an order is placed by parsing them. That's what Dávid's last paragraph is about.

**Comment by DAud-IcI** [comment]
Oliver is right, you only save the base SKU. The variant SKU should be generated similarly to [how the price variant labels are generated](https://github.com/OrchardCMS/OrchardCore.Commerce/blob/main/src/Modules/OrchardCore.Commerce/Drivers/PriceVariantsPartDisplayDriver.cs#L82) and then converted to that upper case format I previously described. It will be used for display (e.g. in the invoice or for shipping) or to look up a specific variant (e.g. for inventory management since you'd want to handle each variant's count separately).

**Comment by porgabi** [comment]
Thanks, getting closer to understanding the issue at hand!

> At this point it would be good to convert the base SKU to upper-case during publish just to be clean since it's case-insensitive in the database anyway. Also now dash is an invalid character in the base SKU so that has to be handled.

Couldn't both of these be handled in the `ProductPartDisplayDriver`, converting SKUs to be upper-case here and preventing dashes by adding validation for the field? If I understood correctly, the dashboard editor is the only interface where SKUs can be entered. If so, is there still a need to "chop off the parts of the SKU variable after the first dash" in `ProductService`?

> It will be used for display (e.g. in the invoice or for shipping) or to look up a specific variant (e.g. for inventory management since you'd want to handle each variant's count separately).

Makes sense; just to clarify, these display surfaces are to be implemented in the future and do not yet exist, right?

**Comment by DAud-IcI** [comment]
> Couldn't both of these be handled in the ProductPartDisplayDriver

Sure.

> If so, is there still a need to "chop off the parts of the SKU variable after the first dash" in ProductService?

Yes, the method that just returns the product should be able to accept the base SKU or the variant SKU.

> these display surfaces are to be implemented in the future and do not yet exist, right?

Yes.

**Comment by porgabi** [comment]
Sziasztok Gabi vagyok.

**Comment by porgabi** [comment]
Having added the new methods to **IProductService**, the test-related classes (`DummyProductService` and `FakeProductService`) that are derived from the interface require an implementation of said methods as well. Should these have the proper implementation (just like in ProductService), or should these be throwing a `NotImplementedException`?

**Comment by DAud-IcI** [comment]
You can use `NotSupportedException`, but include some text to make it clear that it's because the tests don't use it.

**Comment by DemeSzabolcs** [comment]
> Also now dash is an invalid character in the base SKU so that has to be handled.

Couldn't we replace the dash with a "more unique" character? I feel like it could be used in the base SKU too. I also have real examples for this, where the user used dashes in the base SKUs.

**Comment by 0liver** [comment]
Where will that character be used? Will it be displayed to the user?

**Comment by 0liver** [comment]
We could one of the non-standard dash characters that no-one should have in their SKUs - ever™. See [here](https://www.compart.com/en/unicode/category/Pd):

![image](https://user-images.githubusercontent.com/164360/194393618-64491cfa-eaca-4e47-9fb7-6d590587603b.png)

**Comment by DAud-IcI** [comment]
Please Oliver, you are giving me a panic attack.

**Comment by 0liver** [comment]
I was waiting for that 🤣

But on a serious note, why not?

**Comment by DAud-IcI** [comment]
Yep, especially if it were to ever enter into the database (e.g. in invoicing) since MS SQL is very prone to "simplifying" characters like that. Also with an future ERP system integration, some of those still don't use Unicode.

Another concern is search. If the SKU is displayed, then product search by SKU is a logical next step. (Probably) works if the SKU is copy-pasted, but not so much if it was written down and then typed in.

Using exotic characters also limits font choice, you will end up with � instead of dash. 😵 

It's probably not the best idea to play Unicode tricks on anything serializable or user-facing. It's easier and cleaner to just ban dashes from the base SKU. As I said above, SKUs should be upper-cased anyway to avoid case-sensitivity discrepancies since the DB is case-insensitive so the SKU character set is already somewhat limited.

**Comment by 0liver** [comment]
Thanks. I like to understand the why's.

As for banning the dash from SKUs - it does seem like a harsh limitation.

### Some non-authoritative examples that promote the dash

1. Shopify mentions the dash in their [first example for SKUs](https://help.shopify.com/en/manual/products/details/sku#example-sku-formats).
2. [This page](https://www.veeqo.com/blog/what-is-a-sku) uses dashes all over the place.
3. [This e-commerce site](https://help.amplifier.com/en/articles/1804316-what-is-a-sku-and-how-can-i-format-them) says SKUs should be "composed of letters, numbers, underscores, or hyphens".

### Counterexamples

[This page](https://squareup.com/us/en/townsquare/stock-keeping-unit) mentions that SKUs are "usually eight alphanumeric digits" which would exclude the hyphen.

### TL;DR;

Let's choose some other ASCII character.
1. `*`
2. `|`
3. `/`
4. `^`
5. `+`
6. `'`
7. `:`
8. Choose your own (e.g. [here](https://www.lookuptables.com/text/ascii-table)).

**Comment by DAud-IcI** [comment]
Actually the dash example in Shopify is exactly a product with variant information (they just identify it numerically). I say just use underscore if your SKU needs partitioning.
The idea is exactly that the variant SKU should still be a valid SKU, so if SKUs are considered to be "composed of letters, numbers, underscores, or hyphens" then either hyphen or underscore has to be the reserved variant separator.

**Comment by 0liver** [comment]
That drove the point home.

👍🏼

**Comment by Skrypt** [comment]
Why not use one of these chars for tax code applied to a product which would take a precendence over other tax rules when a product is using it. I'm thinking about this because I really don't like the TaxPart. 

Needing to add a Gross Price to a product along with tax code doesn't seem like natural to me. This should be applied automatically based on some rules. One would be to use a SKU that defines the tax code that should be used. This would mean that Tax codes would need to be validated to not include any other chars than what a SKU allows.

Now, my other question is. Why do we disallow using a "dash" on SKU's? Can't we define set of SKU parts with separators? This way, if I want to create a Product that is only one variant I can do PRODUCT-RED+TAXCODE. Knowing that the third element would be the tax code based on separator used?

SKU-VARIANT+TAXCODE
SKU_VARIANT+TAXCODE = generated sku

This way I can create variants of a product as separate content items and apply a rebate on only some of them for example. Then, we don't disallow the "dash" on these product creation because it would be a valid use case?

Or is it that we use the dash for "generated" SKU's as a convention? I would have picked the underscore char as it is less common to use it than a dash for user entries?
